### ISSUE TITLE
Problem with AOT builds in Angular5

### ISSUE THREAD

**Issue by joewoodhouse** [natural-apptitude/ngrx-store-ionic-storage#13]
Hi all,

There seems to be a problem with how the effects work when run in an AOT (`--prod` in ionic) build. Having debugged it, for some reason the `store` is not getting injected into the Effects class properly. So I'm seeing the store be fetched properly, but then when it comes to `this.store.dispatch` I get a runtime error saying store is undefined.

Running in non-aot mode, all works fine as normal.

I'm using:
Ionic: 3.9.2
Angular: 5.0.2
Typescript: 2.4.2
ngrx: 4.1.1

I've found a workaround which is to basically re-implement the effect class inside the application, which seems to work fine. It's just the injection of the Store into this module that seems to be failing.

Maybe someone more knowledgable about the angular-compiler, aot etc can shed some light?

**Comment by okonon** [comment]
I have the same issue.

**Comment by nepsterr** [comment]
Hi. It can be fixed by injecting "Injector" instead of "Store" and use it to "inject" store after. Here's example:
`constructor(private injector: Injector) {
    this.store = this.injector.get(Store);
  }`

**Comment by joewoodhouse** [comment]
@nepsterr For me in --prod/aot mode I just get `injector` as being undefined instead. Seems anything I try to inject is failing. Not sure what's going on. Just updated to latest @ngrx/* and angular 5.0.5 and still no joys.

**Comment by Inlustra** [comment]
@joewoodhouse More of a workaround, but could you not just re-engineer the effects to not use injections? I've used this locally:
```
@Injectable()
export class StorageSyncEffects {

  @Effect() hydrate$: Observable<any> = defer(() =>
    fromPromise(fetchState())
      .map(s => ({
            type: StorageSyncActions.HYDRATED,
            payload: s
        }))
      .catch(e => {
        console.log(`error fetching data from store for hydration`, e);
        return of({
          type: StorageSyncActions.HYDRATED,
          payload: {}
        });
      }));
      
}
```

Might be worth adding a second action in for hydrated error that does nothing but maybe the payload could be the error itself to at least pass the error on to redux-dev-tools or something like that?

**Comment by joewoodhouse** [comment]
Fixed in 4.2.0

**Comment by joewoodhouse** [comment]
@Inlustra thanks for the suggestion! Not sure why it wasn't like that originally anyway, much better code.
